src/relay/pass/type_infer.cc

```diff
@@ -253,14 +231,18 @@ class TypeInferencer : private ExprFunctor<Type(const Expr&)> {
   }
 
   // instantiate the function type with fresh
-  FuncType Instantiate(const FuncTypeNode* fn_ty, Array<Type>* ty_args) {
+  FuncType Instantiate(const FuncTypeNode* fn_ty, Array<Type>* ty_args, const Span& span) {
     tvm::Map<TypeVar, Type> subst_map;
 
     // Build a subsitituion map up from the function type and type arguments.
     // Eventually allow the type vars to be passed in.
-    for (auto ty_param : fn_ty->type_params) {
+    for (size_t i = 0; i < fn_ty->type_params.size(); i++) {
+      auto ty_param = fn_ty->type_params[i];
       IncompleteType fresh = IncompleteTypeNode::make(ty_param->kind);
       subst_map.Set(ty_param, fresh);
+      if (i < ty_args->size()) {
```

Comment:
Why isn't this an invariant?

Reply:
I wasn't sure how permissive we wanted to be with explicit type params. Namely, would a user always have to give _all_ the type params if he or she wanted to specify _any_ of the type params? I'm not sure this is a particularly good way of doing it, so we could make that a requirement to either specify all of them or none of them